### PR TITLE
Refactor CLI

### PR DESCRIPTION
**Pull request by BarkingBad**

### CODE REVIEW

**Comment by BarkingBad** [comment]
Suprisingly it works for dokka-example (note that some paths are absolute, if you want to try it yourself)

```
java -jar dokka-cli-0.11.0-SNAPSHOT.jar -outputDir dokka-cli -format html -generateIndexPages -pluginsClasspath "./mathjax-plugin-0.11.0-SNAPSHOT.jar;./kotlinx-html.jar;/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-jdk8/1.4-M2-eap-70/b25fa3ae35f0a9df2cca793924ef279f0f391b9f/kotlin-stdlib-jdk8-1.4-M2-eap-70.jar;/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-jdk7/1.4-M2-eap-70/404d916c7008edbde9bb72d11b2335c4d7d70727/kotlin-stdlib-jdk7-1.4-M2-eap-70.jar;/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib/1.4-M2-eap-70/9760ac6742c69f9b038dbeab78f7692ba4628fa3/kotlin-stdlib-1.4-M2-eap-70.jar;/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-common/1.4-M2-eap-70/7ca1f6a0923f7deaeea3dc65497990e075eeb5a5/kotlin-stdlib-common-1.4-M2-eap-70.jar;/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains/annotations/13.0/919f0dfe192fb4e063e7dacadee7f8bb9a2672a9/annotations-13.0.jar" -pluginsConfiguration "org.jetbrains.dokka.mathjax.MathjaxPlugin={\"property1\": 1, \"property2\": \"new\", \"property3\": [1, 2, 3]}" -pass "-module example -sourceSetName jvmMain -srcLink src/jvmMain/kotlin=https://github.com/kamilok1965/dokka-example/tree/master/src/jvmMain/kotlin#L -classpath /home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib/1.4-M2-eap-70/9760ac6742c69f9b038dbeab78f7692ba4628fa3/kotlin-stdlib-1.4-M2-eap-70.jar;/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-common/1.4-M2-eap-70/7ca1f6a0923f7deaeea3dc65497990e075eeb5a5/kotlin-stdlib-common-1.4-M2-eap-70.jar;/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains/annotations/13.0/919f0dfe192fb4e063e7dacadee7f8bb9a2672a9/annotations-13.0.jar -analysisPlatform jvm -targets jvm -src src/jvmMain/kotlin -dependentSourceRoots src/commonMain/kotlin;src/jvmAndJsSecondCommonMain/kotlin -samples src/jsMain/resources/Samples.kt -includes src/jvmMain/resources/doc.md" -pass "-module example -sourceSetName jvmAndJsSecondCommonMain -analysisPlatform jvm" -pass "-module example -sourceSetName linuxX64Main -classpath /home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-common/1.4-M2-eap-70/7ca1f6a0923f7deaeea3dc65497990e075eeb5a5/kotlin-stdlib-common-1.4-M2-eap-70.jar -analysisPlatform native -targets linuxX64 -src src/linuxX64Main/kotlin -dependentSourceRoots src/commonMain/kotlin -srcLink src/linuxX64Main/kotlin=https://github.com/kamilok1965/dokka-example/tree/master/src/linuxX64Main/kotlin#L" -pass "-module example -sourceSetName metadata -analysisPlatform common -src src/commonMain/kotlin -samples src/jsMain/resources/Samples.kt" -pass "-module example -sourceSetName jsMain -classpath /home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-js/1.4-M2-eap-70/18134a2693c08f2be3bad8aa4f6ef00da6b938c3/kotlin-stdlib-js-1.4-M2-eap-70.jar;/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-common/1.4-M2-eap-70/7ca1f6a0923f7deaeea3dc65497990e075eeb5a5/kotlin-stdlib-common-1.4-M2-eap-70.jar -analysisPlatform js -targets js -src src/jsMain/kotlin -dependentSourceRoots src/commonMain/kotlin;src/jvmAndJsSecondCommonMain/kotlin -samples src/jsMain/resources/Samples.kt -includes src/jsMain/resources/doc.md -srcLink src/jsMain/kotlin=https://github.com/kamilok1965/dokka-example/tree/master/src/jsMain/kotlin#L"
```

File: runners/cli/src/main/kotlin/cli/main.kt [hunk 1]
```diff
@@ -1,339 +1,344 @@
...
+
+    override val outputDir by parser.option(ArgType.String, description = "Output directory path").required()
 
-    override val format: String by parser.stringOption(
-        listOf("-format"),
-        "Output format (text, html, gfm, jekyll, kotlin-website)",
-        ""
+    override val format by parser.option(
+        ArgType.String,
+        description = "Output format (text, html, gfm, jekyll, kotlin-website)"
```
Comment (kamilok1965) [inline runners/cli/src/main/kotlin/cli/main.kt:21]: We only have `html`, `gfm` and `jekyll` afaik

File: runners/cli/src/main/kotlin/cli/main.kt [hunk 2]
```diff
@@ -1,339 +1,344 @@
...
+    ).default(emptyMap())
+
+    override val pluginsClasspath by parser.option(
+        ArgTypeFile,
+        description = "List of jars with dokka plugins (allows many paths separated by the semicolon `;`)"
+    ).delimiter(";").also {
         Paths.get("./dokka-base.jar").toAbsolutePath().normalize().run {
-            if (Files.exists(this)) it.value.add(this.toFile())
+            if (Files.exists(this)) (it.value as MutableList<File>).add(this.toFile())
             else throw FileNotFoundException("Dokka base plugin is not found! Make sure you placed 'dokka-base.jar' containing base plugin along the cli jar file")
```
Comment (kamilok1965) [inline runners/cli/src/main/kotlin/cli/main.kt:51]: I think we shouldn't throw an exception there as dokka-base is an optional jar. We should log it though

File: runners/cli/src/main/kotlin/cli/main.kt [hunk 3]
```diff
@@ -1,339 +1,344 @@
...
-            entry(configuration)
-    }
 }
 
-
-
+fun main(args: Array<String>) {
+    val configuration = GlobalArguments(args)
+    val generator = DokkaGenerator(configuration, DokkaConsoleLogger)
+    generator.generate()
```
Comment (kamilok1965) [inline runners/cli/src/main/kotlin/cli/main.kt:343]:
```suggestion
    DokkaGenerator(configuration, DokkaConsoleLogger).generate()
```

**Comment by Kordyjan** [comment]
Before closing, create issue about loading plugins options from json file.

File: runners/cli/src/main/kotlin/cli/main.kt (hunk 1 above)
Comment (BarkingBad) [reply runners/cli/src/main/kotlin/cli/main.kt:21]: Salvaged from old one, but will fix

File: runners/cli/src/main/kotlin/cli/main.kt (hunk 2 above)
Comment (BarkingBad) [reply runners/cli/src/main/kotlin/cli/main.kt:51]: There has been a task few months ago to link dokka-base to gradle/maven and enforce to provide jar in CLI, but can chagne

File: runners/cli/src/main/kotlin/cli/main.kt (hunk 3 above)
Comment (BarkingBad) [reply runners/cli/src/main/kotlin/cli/main.kt:343]: Will change

**Comment by BarkingBad** [comment]
Usage: java -jar <dokka_cli.jar> <path_to_config.json>
Configuration for example
```json
{
  "outputDir": "/home/aratajczak/Dokumenty/dokka-example/dokka",
  "format": "html",
  "generateIndexPages": true,
  "impliedPlatforms": [],
  "passesConfigurations": [
    {
      "classpath": [],
      "moduleName": "example",
      "sourceSetName": "commonMain",
      "sourceRoots": [
        {
          "path": "/home/aratajczak/Dokumenty/dokka-example/src/commonMain/kotlin"
        }
      ],
      "dependentSourceRoots": [],
      "dependentSourceSets": [],
      "samples": [],
      "includes": [],
      "includeNonPublic": false,
      "includeRootPackage": false,
      "reportUndocumented": false,
      "skipEmptyPackages": false,
      "skipDeprecated": false,
      "jdkVersion": 8,
      "sourceLinks": [],
      "perPackageOptions": [],
      "externalDocumentationLinks": [],
      "noStdlibLink": false,
      "noJdkLink": false,
      "noAndroidSdkLink": false,
      "suppressedFiles": [],
      "collectInheritedExtensionsFromLibraries": false,
      "analysisPlatform": "common",
      "platform": "metadata",
      "targets": []
    },
    {
      "classpath": [
        "/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-js/1.4-M2-eap-70/18134a2693c08f2be3bad8aa4f6ef00da6b938c3/kotlin-stdlib-js-1.4-M2-eap-70.jar",
        "/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-common/1.4-M2-eap-70/7ca1f6a0923f7deaeea3dc65497990e075eeb5a5/kotlin-stdlib-common-1.4-M2-eap-70.jar"
      ],
      "moduleName": "example",
      "sourceSetName": "jsMain",
      "sourceRoots": [
        {
          "path": "/home/aratajczak/Dokumenty/dokka-example/src/jsMain/kotlin"
        }
      ],
      "dependentSourceRoots": [
        {
          "path": "/home/aratajczak/Dokumenty/dokka-example/src/commonMain/kotlin"
        },
        {
          "path": "/home/aratajczak/Dokumenty/dokka-example/src/jvmAndJsSecondCommonMain/kotlin"
        }
      ],
      "dependentSourceSets": [
        "commonMain",
        "jvmAndJsSecondCommonMain"
      ],
      "samples": [
        "/home/aratajczak/Dokumenty/dokka-example/src/jsMain/resources/Samples.kt"
      ],
      "includes": [
        "/home/aratajczak/Dokumenty/dokka-example/src/jsMain/resources/doc.md"
      ],
      "includeNonPublic": false,
      "includeRootPackage": false,
      "reportUndocumented": false,
      "skipEmptyPackages": false,
      "skipDeprecated": false,
      "jdkVersion": 8,
      "sourceLinks": [
        {
          "path": "src/jsMain/kotlin",
          "url": "https://github.com/kamilok1965/dokka-example/tree/master/src/jsMain/kotlin",
          "lineSuffix": "#L"
        }
      ],
      "perPackageOptions": [],
      "externalDocumentationLinks": [],
      "noStdlibLink": false,
      "noJdkLink": false,
      "noAndroidSdkLink": false,
      "suppressedFiles": [],
      "collectInheritedExtensionsFromLibraries": false,
      "analysisPlatform": "js",
      "platform": "js",
      "targets": []
    },
    {
      "classpath": [],
      "moduleName": "example",
      "sourceSetName": "jvmAndJsSecondCommon",
      "sourceRoots": [],
      "dependentSourceRoots": [],
      "dependentSourceSets": [],
      "samples": [],
      "includes": [],
      "includeNonPublic": false,
      "includeRootPackage": false,
      "reportUndocumented": false,
      "skipEmptyPackages": false,
      "skipDeprecated": false,
      "jdkVersion": 8,
      "sourceLinks": [],
      "perPackageOptions": [],
      "externalDocumentationLinks": [],
      "noStdlibLink": false,
      "noJdkLink": false,
      "noAndroidSdkLink": false,
      "suppressedFiles": [],
      "collectInheritedExtensionsFromLibraries": false,
      "analysisPlatform": "jvm",
      "targets": []
    },
    {
      "classpath": [
        "/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib/1.4-M2-eap-70/9760ac6742c69f9b038dbeab78f7692ba4628fa3/kotlin-stdlib-1.4-M2-eap-70.jar",
        "/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-common/1.4-M2-eap-70/7ca1f6a0923f7deaeea3dc65497990e075eeb5a5/kotlin-stdlib-common-1.4-M2-eap-70.jar",
        "/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains/annotations/13.0/919f0dfe192fb4e063e7dacadee7f8bb9a2672a9/annotations-13.0.jar"
      ],
      "moduleName": "example",
      "sourceSetName": "jvmMain",
      "sourceRoots": [
        {
          "path": "/home/aratajczak/Dokumenty/dokka-example/src/jvmMain/java"
        },
        {
          "path": "/home/aratajczak/Dokumenty/dokka-example/src/jvmMain/kotlin"
        }
      ],
      "dependentSourceRoots": [
        {
          "path": "/home/aratajczak/Dokumenty/dokka-example/src/commonMain/kotlin"
        },
        {
          "path": "/home/aratajczak/Dokumenty/dokka-example/src/jvmAndJsSecondCommonMain/kotlin"
        }
      ],
      "dependentSourceSets": [
        "commonMain",
        "jvmAndJsSecondCommonMain"
      ],
      "samples": [
        "/home/aratajczak/Dokumenty/dokka-example/src/jsMain/resources/Samples.kt"
      ],
      "includes": [
        "/home/aratajczak/Dokumenty/dokka-example/src/jvmMain/resources/doc.md"
      ],
      "includeNonPublic": false,
      "includeRootPackage": false,
      "reportUndocumented": false,
      "skipEmptyPackages": false,
      "skipDeprecated": false,
      "jdkVersion": 8,
      "sourceLinks": [
        {
          "path": "src/jvmMain/kotlin",
          "url": "https://github.com/kamilok1965/dokka-example/tree/master/src/jvmMain/kotlin",
          "lineSuffix": "#L"
        }
      ],
      "perPackageOptions": [],
      "externalDocumentationLinks": [],
      "noStdlibLink": false,
      "noJdkLink": false,
      "noAndroidSdkLink": false,
      "suppressedFiles": [],
      "collectInheritedExtensionsFromLibraries": false,
      "analysisPlatform": "jvm",
      "platform": "jvm",
      "targets": []
    },
    {
      "classpath": [
        "/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-common/1.4-M2-eap-70/7ca1f6a0923f7deaeea3dc65497990e075eeb5a5/kotlin-stdlib-common-1.4-M2-eap-70.jar"
      ],
      "moduleName": "example",
      "sourceSetName": "linuxX64Main",
      "sourceRoots": [
        {
          "path": "/home/aratajczak/Dokumenty/dokka-example/src/linuxX64Main/kotlin"
        }
      ],
      "dependentSourceRoots": [
        {
          "path": "/home/aratajczak/Dokumenty/dokka-example/src/commonMain/kotlin"
        }
      ],
      "dependentSourceSets": [
        "commonMain"
      ],
      "samples": [],
      "includes": [],
      "includeNonPublic": false,
      "includeRootPackage": false,
      "reportUndocumented": false,
      "skipEmptyPackages": false,
      "skipDeprecated": false,
      "jdkVersion": 8,
      "sourceLinks": [
        {
          "path": "src/linuxX64Main/kotlin",
          "url": "https://github.com/kamilok1965/dokka-example/tree/master/src/linuxX64Main/kotlin",
          "lineSuffix": "#L"
        }
      ],
      "perPackageOptions": [],
      "externalDocumentationLinks": [],
      "noStdlibLink": false,
      "noJdkLink": false,
      "noAndroidSdkLink": false,
      "suppressedFiles": [],
      "collectInheritedExtensionsFromLibraries": false,
      "analysisPlatform": "native",
      "platform": "native",
      "targets": []
    }
  ],
  "pluginsClasspath": [
    {
      "path": "/home/aratajczak/.m2/repository/org/jetbrains/dokka/dokka-base/0.11.0-SNAPSHOT/dokka-base-0.11.0-SNAPSHOT.jar"
    },
    {
      "path": "/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-jdk8/1.4-M2-eap-70/b25fa3ae35f0a9df2cca793924ef279f0f391b9f/kotlin-stdlib-jdk8-1.4-M2-eap-70.jar"
    },
    {
      "path": "/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jsoup/jsoup/1.12.1/55819a28fc834c2f2bcf4dcdb278524dc3cf088f/jsoup-1.12.1.jar"
    },
    {
      "path": "/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlinx/kotlinx-html-jvm/0.6.10/adfa3b599d9863ee2d2382ae93abb5621244e8c1/kotlinx-html-jvm-0.6.10.jar"
    },
    {
      "path": "/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-jdk7/1.4-M2-eap-70/404d916c7008edbde9bb72d11b2335c4d7d70727/kotlin-stdlib-jdk7-1.4-M2-eap-70.jar"
    },
    {
      "path": "/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib/1.4-M2-eap-70/9760ac6742c69f9b038dbeab78f7692ba4628fa3/kotlin-stdlib-1.4-M2-eap-70.jar"
    },
    {
      "path": "/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-common/1.4-M2-eap-70/7ca1f6a0923f7deaeea3dc65497990e075eeb5a5/kotlin-stdlib-common-1.4-M2-eap-70.jar"
    },
    {
      "path": "/home/aratajczak/.gradle/caches/modules-2/files-2.1/org.jetbrains/annotations/13.0/919f0dfe192fb4e063e7dacadee7f8bb9a2672a9/annotations-13.0.jar"
    }
  ],
  "pluginsConfiguration": {}
}
```

**Comment by tiembo** [comment]
Is it possible to use the json file but override settings via command line run params?  For example, using most of the settings for the json file but overriding the output directory.

**Comment by BarkingBad** [comment]
It is not supported right now, but we will consider it after release

**Pull request merged**